### PR TITLE
Mutual TLS with ACM

### PR DESCRIPTION
**Pull request by nagmesh**
*Issue #, if available:*
According to our docs, its possible to make mtls work with ACM. This PR has steps to enable that in a test environment
*Description of changes:*
Added a folder and all necessary code.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: walkthroughs/howto-mutual-tls-acm/README.md [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+
+In a basic TLS encryption scenario (for example, when your browser originates an HTTPS connection), the server would present a certificate to any client. In Mutual TLS, both the client and the server present a certificate to each other, and both validate the peer's certificate.
+
+Validation typically involves checking at least that the certificate is signed by a trusted Certificate Authority, and that the certificate is still within its validity period. The application (in our case, Envoy) may also choose to validate other aspects of the certificate, such as the Subject Alternative Name identity signed into the certificate.
+
+In this guide, we will be configuring Envoy proxies using certificates hosted in AWS Secrets Manager, which a modified Envoy image will retrieve during startup. We will have a virtual gateway connected to a single backend service. Both the gateway and backend proxies will present certificates signed by the same Certificate Authority (CA), though you could choose to use separate CAs.
```
Comment (alexbrjo) [inline walkthroughs/howto-mutual-tls-acm/README.md:13]: Can we also mention ACM and the lambda that's used to populate secrets manager here? When does the lambda get called?

File: walkthroughs/howto-mutual-tls-acm/README.md [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+
+We'll start by setting up the basic docker images for our services. All commands will be provided as if run from the same directory as this README.
+
+First, we need to set a few environment variables before provisioning the infrastructure. Please change the value for `AWS_ACCOUNT_ID` and `ENVOY_IMAGE` below.
+
+```bash
+export AWS_ACCOUNT_ID=<your account id>
+export AWS_DEFAULT_REGION=us-west-2
+export ENVIRONMENT_NAME=mtls-ec2-appmesh-example
+export MESH_NAME=mtls-ec2-appmesh
```
Comment (alexbrjo) [inline walkthroughs/howto-mutual-tls-acm/README.md:33]: is this an EC2 only solution or does is also work in FARGATE?

File: walkthroughs/howto-mutual-tls-acm/README.md [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+
+### Step 2: Export our Custom Envoy Image
+
+Finally, we can build and deploy our custom Envoy image. This container has a `/keys` directory and a custom startup script that will pull the necessary certificates from `AWS Secrets Manager` before starting up the Envoy proxy.
+
+```bash
+./src/customEnvoyImage/deploy.sh
+```
+
+> Note: This walkthrough uses this custom Envoy image for illustration purposes; using this method, rotating certificates will require a deployment of your tasks. You may instead choose to mount to your tasks an Elastic Block Storage or Elastic File System volume to store your TLS materials and reference them through the API accordingly.
```
Comment (alexbrjo) [inline walkthroughs/howto-mutual-tls-acm/README.md:61]: > ...using this method, rotating certificates will require a deployment of your tasks.

This is true regardless of the method you choose. App Mesh does not support an integration with ACM for mTLS at this time. The only way to consume ACM certificates is to export them and treat them like file provided certificates (as you've done in this walkthrough).

Please change to something like this:

> App Mesh does not support an integration with ACM for mTLS at this time. These instructions are for exporting an ACM certificate and using it as a customer file provided certificate. Your certificates will not be automatically updated after renewal, rotating certificates will require a restart or deployment of your tasks...

Lets also put a note at the top of this README.

File: walkthroughs/howto-mutual-tls-acm/infrastructure/deploy.yaml [hunk 4]
```diff
@@ -0,0 +1,892 @@
...
+  mtls: !Equals 
+    - !Ref TlsState
+    - mtls
+  setTls: !Or 
+    - Condition: mtls
+    - Condition: 1wayTls
+
+Resources:
+
+  VPC:
```
Comment (alexbrjo) [inline walkthroughs/howto-mutual-tls-acm/infrastructure/deploy.yaml:90]: nit: We typically divide our CFN stacks into logical groups (network infrastructure, ECR repos, ECS cluster, mesh) so if one resource fails to create the rollback and create time won't be super long

**Review by alexbrjo — CHANGES_REQUESTED** [review]
Hello, thanks for the pull request! App Mesh currently does not support an integration with ACM for mutual TLS. We should make this clear in this example to prevent confusion:

1. Move example to the folder `walkthroughs/howto-mutual-tls-file-provided-by-acm`
2. Re-title README to `Configuring Mutual TLS with File Provided TLS Certificates from ACM`

File: walkthroughs/howto-mutual-tls-acm/README.md (hunk 1 above)
Comment (nagmesh) [reply walkthroughs/howto-mutual-tls-acm/README.md:13]: Lambda gets called after ACM is created as a custom resource in the template file
https://github.com/aws/aws-app-mesh-examples/blob/54acdd1b3d8e3d89efc4e530b9527226f278cd5d/walkthroughs/howto-mutual-tls-acm/infrastructure/deploy.yaml#L524-L605

File: walkthroughs/howto-mutual-tls-acm/README.md (hunk 2 above)
Comment (nagmesh) [reply walkthroughs/howto-mutual-tls-acm/README.md:33]: This solution is for ECS Fargate but can run on any compute that supports containers like AWS ECS on EC2, AWS EKS on EC2 or AWS EKS on Fargate

File: walkthroughs/howto-mutual-tls-acm/infrastructure/deploy.yaml (hunk 4 above)
Comment (nagmesh) [reply walkthroughs/howto-mutual-tls-acm/infrastructure/deploy.yaml:90]: Sure. Based on other walkthroughs I can change this to:
- network infrastructure
- ACM
- ECR repos
- ECS cluster
- mesh
- ECS Service

File: walkthroughs/howto-mutual-tls-acm/README.md (hunk 3 above)
Comment (nagmesh) [reply walkthroughs/howto-mutual-tls-acm/README.md:61]: CloudWatch has an event trigger for notifying users on expiration of certs. That can be used to trigger an automation to update the service with `force-new-deployment` option.

File: walkthroughs/howto-mutual-tls-acm/README.md (hunk 1 above)
Comment (nagmesh) [reply walkthroughs/howto-mutual-tls-acm/README.md]: I have add this in the ACM template and mentioned in the README

**Review by nagmesh — COMMENTED** [review]
Added new files for deployment
Changed README and directory name
Tested and verified changes in personal sandbox

File: walkthroughs/howto-mutual-tls-file-provided-by-acm/infrastructure/ecs-service.yaml [hunk 5]
```diff
@@ -0,0 +1,412 @@
...
+            - Name: ENVOY_LOG_LEVEL
+              Value: warning
+            
+          # Set secret environment variable for the container
+          Secrets:
+            - Name: CertSecret
+              ValueFrom: {'Fn::ImportValue': !Sub "${EnvironmentName}:SecretCertArn"}
+
+  GatewayService:
+    Type: 'AWS::ECS::Service'
```
Comment (alexbrjo) [inline walkthroughs/howto-mutual-tls-file-provided-by-acm/infrastructure/ecs-service.yaml:217]: you need a `DependsOn: PublicLoadBalancer` here. or else you can get an exception: "Target group 'WebTargetGroup' doesn't have an associated load balancer"

File: walkthroughs/howto-mutual-tls-file-provided-by-acm/README.md [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+% curl $COLORAPP_ENDPOINT
+yellow
+```
+
+From the bastion, take a look at SSL stats:
+
+```bash
+% curl -s colorteller.mtls.svc.cluster.local:9901/stats | grep -E 'ssl.handshake|ssl.no_certificate'
+listener.0.0.0.0_15000.ssl.handshake: 1
+listener.0.0.0.0_15000.ssl.no_certificate: 0
```
Comment (alexbrjo) [inline walkthroughs/howto-mutual-tls-file-provided-by-acm/README.md:335]: So I'm getting incosistent behavior here. First time I run the command I get this output but the next time I get 9/8. Like below:

```
sh-4.2$ curl -s colorteller.mtls.svc.cluster.local:9901/stats | grep -E 'ssl.handshake|ssl.no_certificate'
listener.0.0.0.0_15000.ssl.handshake: 9
listener.0.0.0.0_15000.ssl.no_certificate: 8
sh-4.2$ curl -s colorteller.mtls.svc.cluster.local:9901/stats | grep -E 'ssl.handshake|ssl.no_certificate'
listener.0.0.0.0_15000.ssl.handshake: 1
listener.0.0.0.0_15000.ssl.no_certificate: 0
sh-4.2$ curl -s colorteller.mtls.svc.cluster.local:9901/stats | grep -E 'ssl.handshake|ssl.no_certificate'
listener.0.0.0.0_15000.ssl.handshake: 9
listener.0.0.0.0_15000.ssl.no_certificate: 8
sh-4.2$ curl -s colorteller.mtls.svc.cluster.local:9901/stats | grep -E 'ssl.handshake|ssl.no_certificate'
listener.0.0.0.0_15000.ssl.handshake: 1
listener.0.0.0.0_15000.ssl.no_certificate: 0
```

It seems to toggle back and forth. Can you explain this? / is this intended?

**Review by alexbrjo — APPROVED** [review]
Looks good! I was able to successfully run the walkthrough (with the addition of that `dependsOn` Cloud Formation line).

File: walkthroughs/howto-mutual-tls-file-provided-by-acm/README.md (hunk 6 above)
Comment (alexbrjo) [reply walkthroughs/howto-mutual-tls-file-provided-by-acm/README.md:335]: I didn't wait for the deployment to complete. These requests were being load balanced over the tls and mtls deployments. Looks good!

File: walkthroughs/howto-mutual-tls-file-provided-by-acm/infrastructure/ecs-service.yaml (hunk 5 above)
Comment (nagmesh) [reply walkthroughs/howto-mutual-tls-file-provided-by-acm/infrastructure/ecs-service.yaml:217]: I added the `DependsOn` to the WebTargetGroup directly
https://github.com/aws/aws-app-mesh-examples/blob/b32767f5c4b0a91912fe9b9e3486e01b6b39fc6e/walkthroughs/howto-mutual-tls-file-provided-by-acm/infrastructure/ecs-service.yaml#L255-L256

**Pull request merged**